### ISSUE TITLE
question: How to use automatic union types?

### ISSUE THREAD

**Issue by wu-clan** [DanCardin/cappa#74]
Can you add more clarification on the processing logic for union types? The current results don't seem to make sense for union types (documents)

MRE:
```python
from __future__ import annotations

from dataclasses import dataclass
from typing import Annotated, List, Tuple

import cappa


@dataclass
class ArgTest:
    arg: Annotated[float | int | bool | str | List[str], cappa.Arg(short=True, long=True)]
    foo: Annotated[bool | Tuple[str, ...], cappa.Arg(short=True, long=True, num_args=-1)]


a = cappa.parse(ArgTest, argv=['--arg', '3.14', '--foo'])
print(a)
# ArgTest(arg=['3', '.', '1', '4'], foo=())

b = cappa.parse(ArgTest, argv=['--arg', '1', '--foo'])
print(b)
# ArgTest(arg=['1'], foo=())

# c = cappa.parse(ArgTest, argv=['--arg', '--foo'])
# print(c)
# Error: Argument requires 1 values, only found 0 ('' so far)

d = cappa.parse(ArgTest, argv=['--arg', 'True', '--foo'])
print(d)
# ArgTest(arg=['T', 'r', 'u', 'e'], foo=())

e = cappa.parse(ArgTest, argv=['--arg', 'True', '-a', 'False', '--foo'])
print(e)
# ArgTest(arg=['F', 'a', 'l', 's', 'e'], foo=())

f = cappa.parse(ArgTest, argv=['-a', 'a', '-f', 'ff', 'f', 'f'])
print(f)
# ArgTest(arg=['a'], foo=('ff', 'f', 'f'))
```

**Comment by pawamoy** [comment]
Looks like it's coalescing in reverse order.

**Comment by wu-clan** [comment]
Thanks for the heads up, I updated the examples.

**Comment by pawamoy** [comment]
I tried reversing the types in the union, to no avail: Cappa keeps coalescing `"3.14"` into `["3", ".", "1", "4"]` :thinking:

**Comment by DanCardin** [comment]
Inference has a precedence order defined among options with things like str/bool/int (since str/bool will always successfully map any value). For things in the same group, the order in the union would matter; but for things among different priorities, you'll always end up with the same inference, regardless of order.

In this case, i think the issue is that `list` and `tuple` fundamentally affect the way the parser parses the input values and map the output values. The current inference code isn't smart enough (and maybe/probably can't be) to determine the right ArgAction from your code and is ending up with the default `set`, so you end up with a string out of the parser. But then at mapping stage it sees `list[str]` and does `list('3.14')`, yielding your result.

I assume what you want done is `ArgAction.append` and then the result value to be mapped to an int/float/bool etc if there's only one value and it's able to be parsed as that type. I cant see that sort of logic making sense generally.

What I assume you probably need/want is:

```python
def parse_unknown(value: Value):
    # do your logic to map the resultant value into the base types yourself

@dataclass
class ArgTest:
    arg: Annotated[float | int | bool | str | List[str], cappa.Arg(short=True, long=True, action=ArgAction.append)]
```

---
Perhaps there should be a ValueError coming out of the usage of the combination of these kinds of incompatible unions, but it's at least not immediately obvious to me how the above definition is supposed to reconcile something like `int | list[str]` that mean different things

**Comment by wu-clan** [comment]
This seems to me to work, and it does work depending on the order in which document union types are processed.

For List/Tuple types, they are always validated in advance, and when you set `ArgAction.append` or `num_args=-1` it can take and return empty arguments, and you can utilize `Arg(parse=custom_parse)` to output the bool type.

Obviously, some incompatible union combinations don't make sense, and in theory, a ValueError or warning is really needed to remind the user how to use them correctly.

**Comment by DanCardin** [comment]
So just to make sure I understand the above comment, with an explicitly set Action/num_args and custom `parse`, you're able to get the behavior you want? (and otherwise is an acceptable/logical solution?)

If so, if were to add a check (Essentially just something that would have errored and maybe indicated what you should have done), that wouldnt immediately break your above solution and have been clear?:
* if action is unset, num_args is unset, and the annotation is a union who's variants include a sequence-type and a non-sequence type, then emit an error something like "Sequence types (list/tuple/set) cannot be unioned with non-sequence types unless `action` or `num_args` set (and likely also `parse`)".

**Comment by wu-clan** [comment]
My previous comments are not entirely valid.

View this MRE.
```python
from __future__ import annotations

from dataclasses import dataclass
from typing import Annotated, List, Tuple

import cappa
from cappa import ArgAction
from cappa.parser import Value


def custom_union_parser(value: Value) -> bool | Value:
    if len(value) == 0:  # type: ignore
        return True
    else:
        return value


@dataclass
class ArgTest:
    arg: Annotated[
        bool | List[str],
        cappa.Arg(short=True, long=True, default=[], parse=custom_union_parser, num_args=-1)
    ]
    foo: Annotated[
        bool | Tuple[str, ...],
        cappa.Arg(short=True, long=True, default=(), parse=custom_union_parser, num_args=-1)
    ]


a = cappa.parse(ArgTest, argv=['--arg', '--foo'])
print(a)
# ArgTest(arg=True, foo=True)
# if arg set action=ArgAction.append, not num_args=-1, then output
# Error: Argument requires 1 values, only found 0 ('' so far)

b = cappa.parse(ArgTest, argv=['--arg', 'True', '--foo', 'False'])
print(b)
# ArgTest(arg=['True'], foo=['False'])

c = cappa.parse(ArgTest, argv=['--arg', 'a1', 'a2', '--foo', 'f1', 'f2'])
print(c)
# ArgTest(arg=['a1', 'a2'], foo=['f1', 'f2'])
```

When using federation, if you use simple custom parsing, action=ArgAction.append and num_args=-1 behavior are not consistent.

Although this can get the results I want, the bool type doesn't seem to make any real sense, just reminding myself that it can return a Bool value.

I can add more unions (int | float |...), but I will continue to customize the parsing to get the expected value, which is still ostensibly List / Tuple.

So when using both a sequence-type and a non-sequence type, it can be considered meaningless? Should the use of these federation types be enforced rules? 
For example (this may not be the standard). 

1. If you define a sequence-type, you cannot use other types at the same time. 
2. `float | int | str` can be used simultaneously. 
3. bool is only used alone. 
......

Union typing is a complex process, and even in pydantic, there are still problems.

**Comment by wu-clan** [comment]
Fully automated parsing of union types is certainly a good thing, but it takes a lot of time and effort. For developers, in cappa, this requires not only input parsing, but also output and actions, so forcing federated types seems easier to implement.

I don't seem to see joint type handling in click / typer / argparse. Usually, a parameter declares only one type and parses input according to that type.

**Comment by DanCardin** [comment]
tl;dr there's some happy medium of extra compatibility checks that need to be added which inform the user that their annotations cant be produced from the `Arg` settings they've configured. The specifics of which are to be determined.

The original issue statement would ideally have yielded an error something like `ValueError("Sequence types (list/tuple/set) cannot be unioned with non-sequence types unless action or num_args are set (see [docs](annotations docs link with more context))")`

And then some additional prose in the docs that explains what's going on here, why, and perhaps some more on the distinction between append/num_args

Otherwise, I dont think there's a bug/feature/enhancement being requested anymore, right?

---

> `a = cappa.parse(ArgTest, argv=['--arg', '--foo'])`
> if arg set action=ArgAction.append, not num_args=-1, then output
> Error: Argument requires 1 values, only found 0 ('' so far)

The options aren't analagous. append allows multiple instances of an option, which happens to yields a list out of the parser, one **item** per instance. num_args controls the number of parser values that are consumed into a single **item**, which may be a list if it's > 1.

Given something like `--foo 1` you'll see:

| action | num_args | result |
|-|-|-|
| set      | 1                | 1         |
| set      | -1              | [1]      |
| append | 1              | [1]      |
| append | -1            | [[1]]    |

You happen to get a list when they're used independently just because of each of their behaviors.

---

Use of `parse` essentially opts you out of type inference/mapping, where you need to implement the reaction to the resultant value you get on your own. If you have complex output mapping requirements that require custom logic to handle, this seems unavoidable.

There are a series of [functions which can be used](https://github.com/DanCardin/cappa/blob/main/src/cappa/annotation.py#L12) to make this easier to compose, perhaps.

---

> 1. If you define a sequence-type, you cannot use other types at the same time.
> 2. float | int | str can be used simultaneously.
> 3. bool is only used alone.

I think bool can be used with other scalar types. In order of preference: None/arbitrary types (like Decimal) > float > int > bool > str. So perhaps bool/str dont make sense together, but otherwise i think the scalars generally **can**.

but yes, i think sequence types and scalar types shouldn't be allowed together.

The only caveats to that are in the rest of the configuration

* `parse`: When set, i dont see why it shouldn't be allowed. the user is opting into arbitrary behavior, they should be allowed to annotate the Arg with the annotation that matches their function's return value
* `num_args=-1` (without parse): require the source types to be sequence types?
* `ArgAction.append` (without parse): require the source types to be sequence types?

**Comment by wu-clan** [comment]
> Otherwise, I dont think there's a bug/feature/enhancement being requested anymore, right?

Yeah, I think that's probably what needs to be done, error alerts and documentation，Otherwise there is no union but just [CLI Parameter Types like Typer](https://typer.tiangolo.com/tutorial/parameter-types/)?

> parse: When set, i dont see why it shouldn't be allowed. the user is opting into arbitrary behavior, they should be allowed to annotate the Arg with the annotation that matches their function's return value
num_args=-1 (without parse): require the source types to be sequence types?
ArgAction.append (without parse): require the source types to be sequence types?

They all seem reasonable.

My only question might be why only `bool | str` might not make sense? The bool type doesn't usually take arguments in the CLI, but other types need to take them, and assuming that bool can be arbitrarily unionized with a type, it will be automatically inferred to be bool when no arguments are entered, and conversely, depending on the order of the None/arbitrary types (like Decimal) > float > int > str
